### PR TITLE
Add RedisSentinel class for score cache

### PR DESCRIPTION
**Pull request by Ladsgroup**
Bug: T210579

### CODE REVIEW

**Comment by Ladsgroup** [comment]
Tested locally and it works.
Make two replicas of redis localhost master (by doing something like redis-server -p 6378 and then "slaveof localhost 6379" on redis-cli -p 6378)
Then make this sentinel.conf
```
sentinel monitor mymaster 127.0.0.1 6379 2
sentinel down-after-milliseconds mymaster 5000
sentinel failover-timeout mymaster 60000
```
"mymaster" is name of the cluster, you can make it 'orescache'.

Then you need to setup three sentinel nodes at ports 5000, 5001, and 5002.
redis-server /path/to/sentinel.conf --sentinel
Then change the config to use sentinel.

File: ores/score_caches/redis_sentinel.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+        return ":".join(str(v) for v in key_values)
+
+    @classmethod
+    def from_parameters(cls, hosts, ttl=None, prefix=None, cluster=None,
+                        socket_timeout=None):
+        try:
+            from redis.sentinel import Sentinel
+        except ImportError:
+            raise ImportError("Could not find redis-py.  This packages is " +
+                              "required when using ores.score_caches.Redis.")
```
Comment (halfak) [inline ores/score_caches/redis_sentinel.py]: "required when using ores.score_caches.RedisSentinel."

Comment (Ladsgroup) [reply ores/score_caches/redis_sentinel.py]: Done

File: ores/score_caches/__init__.py [hunk 2]
```diff
@@ -2,5 +2,6 @@
 from .empty import Empty
 from .lru import LRU
 from .redis import Redis
+from .redis_sentinel import RedisSentinel
```
[the remark this answers is not in the document]
Comment (Ladsgroup) [reply ores/score_caches/__init__.py]: The class basically performs as Redis + Sentinel (Sentinel on its own can't give us anything). I like to keep it that way.

File: ores/score_caches/redis.py [hunk 3]
```diff
@@ -81,3 +84,84 @@ def from_config(cls, config, name, section_key="score_caches"):
...
+
+    def store(self, score, context_name, model_name, rev_id, version=None,
+              injection_cache=None):
+        key = self._generate_key(
+            context_name, model_name, rev_id, version=version,
+            injection_cache=injection_cache)
+
+        master = self.sentinel.master_for(self.cluster, socket_timeout=self.socket_timeout)
+        sentinel_logger.debug("Storing score at {0} in master".format(key))
+        master.setex(key, self.ttl, bytes(json.dumps(score), 'utf-8'))
```
Comment (adamwight) [inline ores/score_caches/redis.py]: For example, Redis and RedisSentinel could vary only here, so `store` might be shared but `_setex` would be overridden in the subclass.

File: ores/score_caches/redis.py [hunk 4]
```diff
@@ -81,3 +84,84 @@ def from_config(cls, config, name, section_key="score_caches"):
     def hash_cache(cls, cache):
         sorted_tuple = tuple(sorted(cache.items()))
         return sha1(bytes(str(sorted_tuple), 'utf8')).hexdigest()
+
+
+class RedisSentinel(ScoreCache):
```
[the remark this answers is not in the document]
Comment (Ladsgroup) [reply ores/score_caches/redis.py]: I tried it and failed miserably, that reason being is that Redis requires "redis" argument and attribute but that doesn't exist in RedisSentinel and we can't instantiate the class.

File: ores/score_caches/redis.py [hunk 5]
```diff
@@ -5,9 +5,12 @@
 from .score_cache import ScoreCache
 
 logger = logging.getLogger("ores.score_caches.redis")
+sentinel_logger = logging.getLogger('ores.score_caches.redis_sentinel')
```
[the remark this answers is not in the document]
Comment (Ladsgroup) [reply ores/score_caches/redis.py]: hmm, yeah I think so. Let me do it.

[the remark this answers is not in the document]
Comment (Ladsgroup) [reply ores/score_caches/redis.py]: I do it for redis sentinel but I think the line 7 also needs fixing. Will get to it later.

File: ores/score_caches/redis.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (adamwight) [reply ores/score_caches/redis.py]: Then they can both include a third class?

If you don't think this is worthwhile, I'm fine.

[the remark this answers is not in the document]
Comment (adamwight) [reply ores/score_caches/redis.py]: Possibly, the difference between classes is as simple as a new `self.get_redis(WRITE_MODE)` which returns `self.sentinel.master_for` or `self.redis` according to subclass implementation.

**Comment by adamwight** [comment]
We can live with the duplication for now—merging.

**Pull request merged**